library_registry/admin/controller.py

```diff
@@ -18,33 +23,66 @@
     Validation,
 )
 
+
 class ViewController(BaseController):
     def __call__(self):
-        username = session.get('username', '')
+        """View controller for setting Flask Session or Flask JWT to verify user
+
+        Returns:
+            Response: rednered template string with user identity
+        """
+        if verify_jwt_in_request(optional=True):
+            username = get_jwt_identity()
+        else:
+            username = session.get('username', '')
+
         response = Response(render_template_string(
             admin_template,
             username=username
         ))
         return response
 
+
 class AdminController(BaseController):
 
     def __init__(self, app, emailer_class=Emailer):
         super(AdminController, self).__init__(app)
         self.emailer = emailer_class
-    
-    def log_in(self):
+
+    def log_in(self, jwt_preferred=False):
+        """End point for login with requesting flask sessions or JWT tokens
+
+        Returns:
+            err: Invalid credentials
+            flask session: flask session and redirect
+            jwt tokens: JWT tokens as Set-Cookie headers and redirect
+        """
         username = request.form.get("username")
         password = request.form.get("password")
-        if Admin.authenticate(self._db, username, password):
+        if not Admin.authenticate(self._db, username, password):
+            return INVALID_CREDENTIALS
+        if not jwt_preferred:
             session["username"] = username
             return redirect(url_for('admin.admin_view'))
-        else:
-            return INVALID_CREDENTIALS
+        access_token = create_access_token(identity=username)
+        response = make_response(
+            redirect(url_for('admin.admin_view')), 302)
+        set_access_cookies(response, access_token)
```

Comment:
wait, I'm afraid I've led us down a bad path here. When we de-couple the admin interface and deploy it as a standalone app this redirect won't function anymore, correct? We _would_ need another endpoint to directly return the token in that case if we need to support redirects and an explicit token return flow.

Reply:
I'm not sure I'm following so let me know if I haven't addressed what you mean. We are using `redirect` here for the Response object it returns to help build a larger response in `make_response`. I'm not sure if `return redirect(url_for('admin.admin_view'))` would stop working since it just returns a response. It certainly wouldn't do what was expected for login because there'd be no cookie sent to the client if that's what you mean. That's why we're explicitly building the larger response instead with the 302 and the cookie.

I don't think we need to support redirects, but we should have that conversation with frontend to make sure they are prepared to handle them. It seems there shouldn't be too much of a change from how they read 302s now. You are right that we will need another `/refresh` endpoint when we implement the explicit token return flow (the current cookies implementation uses the implicit refresh process to behave in a more expected way), and the logic for hitting that refresh endpoint has to be implemented on the frontend.

Reply:
If we remove redirects from our application flow, then I think the best option that can plug and play with both library_reg and CM is Explicit cookies.  However, with the library_reg I think that we are only having to support web browsers for the protected admin route.  If this is the case, and we aren't planning a massive expansion of this system, implicit cookies work for them very automagically.  Even if we decide to pull redirects, we just need to pass a Response object with set_cookie called on it.

Reply:
The backend cannot handle redirects directly, but the backend is still in charge of the response status codes and any relevant URLS that need to be returned. You can't eliminate sending proper response codes. I think there is some confusion here around what it means to handle redirects vs the forming of responses, and that confusion may be mine. Our API can't handle redirects and absolutely shouldn't even try. It absolutely should return 302s and URLs when relevant.

By explicit cookies, do you mean explicit refresh? Explicit refresh is the recommended approach for supporting headers, which is originally the only thing we were going to support. Implicit refresh is the recommended approach for cookies, which is what we then decided to add support for and focus on first. Explicit refresh requires refresh tokens, and the documentation clearly states that explicit refresh tokens are not recommended for cookies. If you look at the setup of the code for explicit refresh for cookies, it becomes clear in my opinion why it's a bad design choice....it doesn't behave like cookies at all. It seems we keep coming back to the idea of explicit refresh with cookies, and I apologize for the changing requirements that may have confused that, but I think we should continue with the approaches that have already been outlined for cookies.

Reply:
At the moment we will leave our current redirects, as this does not need to be changed to implement the extension.  With the explicit tokens the log_in will be expanded and only return a token.  Leaving redirect up to the front end.